### PR TITLE
New behaviour based on sub-themes

### LINKED ISSUE #1196: New behavior based on sub theme
Already today sub themes are delivered by pyramid_oereb.

They are handled in several places but are never really processed. Only some sorting mechanism is constructed on it.

In the moment all related stuff is named "SubTheme" and not "SubTopic" in this project. So to find usages this is most important to know because it is often called sub topic in daily use.

The sub theme was a free text in the old specification but is of type "ThemeCode" now. This means, every entry in Data has to fit the pattern described in the specification of the [DATA-Extract](https://github.com/openoereb/pyramid_oereb/files/6442562/OEREB-Data-Extract_de.pdf) appendix A and B.

### What needs to be done

It is important, that sub themes should only be used in the theme "land use plans". It's still not decided if this leads to a special fall back behavior when other themes contains sub themes.

**Datawise:**
To fit the needs, data must contain new content for the field sub theme which fits the specified rules. This is nothing to develop but must stay in mind when new version is in use some day.

**Code**
In the future approach pyramid_oereb must produce topics on runtime based on data content. This means:

1. before some calculation takes place we need to iterate over the data which might be affected
2. group data into n groups depending on the field subtheme
3. create new topics based on the groups we have
4. all calculation (area, length, point count) can be done like before

All runtime generated topics need to get dedicated titles and texts so that they can be recognized as a special topic of a given one. Probably the best place to look at to get a sense what this might be, is the [specification of the static extract](https://github.com/openoereb/pyramid_oereb/files/6442522/Weisung-statischer-Auszug-de.pdf). But it affects the complete pyramid_oereb, not only the print.

All texts should be taken out of the provided [models](http://models.geo.admin.ch/V_D/OeREB/OeREBKRM_V2_0_Texte.xml).

### PR DESCRIPTION
**Pull request by marionb**
closes #1196 
TODO:
- [x] implement new sub-themes on the level of themes
- [x] add test data
- [x] fix json & xml extracts
- [x] Fix Print
- [x] implement sorting for themes and sub-themes
- [x] check if this sorting is what is expected
- [x] fix current tests
- [ ] add tests if needed

@vvmruder and @voisardf Currently I am sorting the themes and sub-themes according to their `extract_index` and no longer with what was implemented earlier. Before I clean out the old sort mechanism can you tell me if the is fine like this or if we need something like the old sorting mechanism again?
Also the sorting is now done in the extract together with the sorting of the `law status`, so that we have to sort this part only once.

### CODE REVIEW

**Comment by voisardf** [comment]
Yes, sorting by the given index is the new way it should work.

Sorting principles are described on page 8 ff of the static extract specs: https://github.com/openoereb/pyramid_oereb/files/6442562/OEREB-Data-Extract_de.pdf and detailed on page 13, first line.
Interestingly enough, the sortindex sems not to appear in the extract model and from the specifications it is not exactly clear if it should be passed as integer or string: https://github.com/openoereb/models/issues/3

**Comment by marionb** [comment]
thanks @voisardf I removed the old sorter logic.
Currently I implemented the index thinking it was an integer (I think it will also work wit a string). I will leave it up to the person who integrates the data to set the index correctly so that the sorting is in accordance with the spec of the the static extract.

**Review by vuilleumierc — COMMENTED** [review]
Tested locally and it works well, thanks for putting all this together!
Great that you added new geometries (points and lines with the new LandUsePlans PLRs), although I am surprised that in the static extract we only ever see the same rectangle, although "Anteil" has sometimes different units (m^2, m, -). Do you know why that is? I don't think it is related to your PR though as it as always been the case.
Otherwise I just have a few comment in the code

**Comment by voisardf** [comment]
I also tested the branch locally and got the extract (json/xml/pdf). The pdf only when commenting the line 8 regarding pdf conformance PDF/1-a. Which is still an open issue regarding the logos. So that seems fine to me.
I also get seven failing tests. Most of them schema-validation stuff. Do you have the same? @vuilleumierc @marionb

**Comment by vuilleumierc** [comment]
Yes I also get the usual seven failing tests:
```
FAILED tests/contrib/print_proxy/test_mapfish_print.py::test_mapfish_print_entire_extract
FAILED tests/renderer/test_xml.py::test_extract_against_schema[parameter0-test_extract0] - lxml.etree.XMLSchemaParseError: Element '{http://www.w3.org/2001/XMLSchema}import':...
FAILED tests/renderer/test_xml.py::test_extract_against_schema[parameter1-test_extract1] - lxml.etree.XMLSchemaParseError: Element '{http://www.w3.org/2001/XMLSchema}import':...
FAILED tests/renderer/test_xml.py::test_extract_against_schema[parameter2-test_extract2] - lxml.etree.XMLSchemaParseError: Element '{http://www.w3.org/2001/XMLSchema}import':...
FAILED tests/xtf_import/test_legend_entry.py::test_parse - TypeError: 'topic' is an invalid keyword argument for LegendEntry
FAILED tests/xtf_import/test_public_law_restriction.py::test_parse - TypeError: 'legend_text' is an invalid keyword argument for PublicLawRestriction
FAILED tests/xtf_import/test_federal_topic.py::test_load - SystemExit: 1
```

**Comment by marionb** [comment]
@voisardf Yes I also have these test that fail. As far as I saw they also fail on master and have dedicated issues so I do not see is as: in the scope of this PR, to fix them -> https://github.com/openoereb/pyramid_oereb/wiki/DEV-Time-contracts

The PDF that fails it is also correct, unless you remove the 3 lines.

**Comment by vvmruder** [comment]
If it was approved an tested by @voisardf you can merge it. I don't have the time today to test it :)

**Review by vuilleumierc — APPROVED** [review]
Looks very good, thanks!

**Comment by marionb** [comment]
Any one has a clue how to fix this codacity issue? I am stating to think that codacity is actually wrong - or the error is a completely different one than what is stated.

**Review by voisardf — COMMENTED** [review]
Code looks good to me. JSON/XML data extract work, but I could not generate a PDF from this branch. Normal? Or do I have to be  on a particular mpf-branch?

**Comment by marionb** [comment]
@voisardf thanks for testing and reviewing. To generate the PDF you can use the branch https://github.com/openoereb/pyramid_oereb_mfp/tree/Turn_Off_PDFA
The current master does not work because some commits are missing in this PR. Once merging this in to master it will work again with the OEREB_MFP master.

File: pyramid_oereb/standard/templates/plr_integer_primary_keys.py.mako
Comment (marionb) [reply pyramid_oereb/standard/templates/plr_integer_primary_keys.py.mako:214]: I removed the file :-)

File: docker/config.yaml.tmpl
Comment (marionb) [reply docker/config.yaml.tmpl]: Good idea if there is only one canton using this.
I changed it.

**Review by voisardf — APPROVED** [review]
I was not able to get the PDF yet, I still get _Ein oder mehrere ÖREB-Themen stehen momentan nicht zur Verfügung. Daher kann kein Auszug erstellt werden. Versuchen Sie es zu einem späteren Zeitpunkt erneut. Wir entschuldigen uns für die Unannehmlichkeiten._

but if you need the PR to advance, ok for me to merge it while I try to get a better idea where the mfp error lies

**Comment by marionb** [comment]
@voisardf I'll re-base the PR on to master. It will make testing easier.

**Pull request merged**
